### PR TITLE
Add dataset type olmo_grain for AI2 OLMo numpy pretrain mixes

### PR DESCRIPTION
**Pull request by gagika**
# Description

- Adds `dataset_type=olmo_grain`, a Grain-based input pipeline for AI2's
  pre-tokenized OLMo numpy mixes (e.g. [`OLMo-mix-0925-official.txt`](https://github.com/allenai/OLMo-core/blob/main/src/olmo_core/data/mixes/OLMo-mix-0925-official.txt)).
  Reads headerless `.npy` token streams from a gcsfuse mount, applies
  OLMo-core's repeated-n-gram filter, and yields the shapes the MaxText
  pretrain trainer expects.
- Stateless sampler: record at step k is a pure function of
  `(seed, shard, k)`. Resume reads the latest step from
  `config.checkpoint_dir` and shifts the sampler — no Grain iterator
  state in the checkpoint.
- Ships two data tools (`download_olmo_data_to_gcs.py` with HTTP-Range
  resume; `build_olmo_npy_index.py` for header-scan indexing) and two
  launchers (`run_olmo3_7b_grain_smoke.sh`,
  `run_olmo3_7b_grain_resume_test.sh`).

# Tests
- Unit tests pass (`tests/unit/input_pipeline/olmo_*`)
- Smoke train: 50 steps, loss 11.99 → 8.93 on v4-8 (4-layer bf16)
- Resume test: Run B picks up at step 50 with loss continuity 8.931 → 8.930

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

authored-by: @aireenmei

### CODE REVIEW

File: tests/unit/input_pipeline/olmo_data_grain_test.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+          idx,
+          seed=0,
+          batch_size=4,
+          shard_index=0,
+          shard_count=1,
+          apply_ngram_filter=False,
+          # Disable shift so we can read the raw ``instance_id`` field for
+          # this audit-style test. (Production runs always use shift=True.)
+          shift_to_inputs_targets=False,
+          grain_worker_count=0,
```
Comment (aireenmei) [inline tests/unit/input_pipeline/olmo_data_grain_test.py]: Does this test aim to test grain_worker_count=2 ?

File: src/maxtext/input_pipeline/olmo_data_grain.py [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+
+import hashlib
+import threading
+from typing import Any, Dict, List, Optional
+
+import numpy as np
+
+# Grain is structured so that the public Python API lives under
+# ``grain.python``. Tests can monkeypatch the ``_grain`` module attribute.
+import grain.python as _grain  # type: ignore[import-not-found]
```
Comment (aireenmei) [inline src/maxtext/input_pipeline/olmo_data_grain.py]: Can just be `import grain`

File: src/maxtext/input_pipeline/olmo_data_grain.py [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+  if shift_to_inputs_targets:
+    ops.append(ShiftToInputsTargets())
+  ops.append(_grain.Batch(batch_size=batch_size, drop_remainder=True))
+
+  # Grain >=0.2.16 expects ``shard_options`` on the DataLoader (sharding used
+  # to live on the Sampler). Even though our sampler already does the
+  # shard-by-rank slicing, Grain still requires this object to validate
+  # checkpoint compatibility.
+  shard_options = _grain.ShardOptions(shard_index=shard_index, shard_count=shard_count, drop_remainder=True)
+  return _grain.DataLoader(
```
Comment (aireenmei) [inline src/maxtext/input_pipeline/olmo_data_grain.py]: grain.DataLoader with sampler and a list of operations is the old API. Now it's recommended to use the new MapDataSource API (https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/input_pipeline/grain_data_processing.py#L94), and use dataset.<method()> to apply transformations: https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/input_pipeline/grain_data_processing.py#L49

File: src/maxtext/input_pipeline/olmo_data_grain.py [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+      arr = np.memmap(path, dtype=dtype, mode="r")
+      self._mmaps[path] = arr
+      while len(self._mmaps) > self._max:
+        # Evict oldest.
+        oldest = next(iter(self._mmaps))
+        del self._mmaps[oldest]
+      return arr
+
+
+class OlmoIndexSampler:
```
Comment (aireenmei) [inline src/maxtext/input_pipeline/olmo_data_grain.py:153]: Why do we need a new Sampler? Can we re-use grain's sampler?

File: tests/unit/input_pipeline/olmo_data_grain_test.py (hunk 1 above)
Comment (gagika) [reply tests/unit/input_pipeline/olmo_data_grain_test.py]: Good catch — fixed, now passing grain_worker_count=2.

File: src/maxtext/input_pipeline/olmo_data_grain.py (hunk 2 above)
Comment (gagika) [reply src/maxtext/input_pipeline/olmo_data_grain.py]: Done. Switched to import grain and updated call sites to the public submodule paths   (grain.sources.RandomAccessDataSource, grain.transforms.Map, grain.transforms.Batch, grain.sharding.ShardOptions).

File: src/maxtext/input_pipeline/olmo_data_grain.py (hunk 3 above)
Comment (gagika) [reply src/maxtext/input_pipeline/olmo_data_grain.py]: Coupled with next comment. The custom sampler powers stateless initial_step resume — pure function of (seed, shard, train_step *  batch_size), no Grain iterator-state checkpoint needed (pinned by olmo_data_grain_resume_test.py).  MapDataset.shuffle().repeat() has no step-offset hook, so migrating would drop that. We can do a follow follow-up fix if you think it's needed.

File: src/maxtext/input_pipeline/olmo_data_grain.py (hunk 4 above)
Comment (gagika) [reply src/maxtext/input_pipeline/olmo_data_grain.py:153]: Same as above comment.

File: docs/guides/data_input_pipeline/olmo_grain.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+from a gcsfuse mount, shards across hosts, optionally masks repeated-n-gram
+instances, and yields the shapes the MaxText pretrain trainer expects.
+
+## Quick start
+
+1. **Download the data** to a GCS bucket:
+
+   ```bash
+   python tools/data_generation/download_olmo_data_to_gcs.py \
+       --mix-file /path/to/OLMo-mix-0925-official.txt \
```
Comment (RissyRan) [inline docs/guides/data_input_pipeline/olmo_grain.md]: Will this path be at GCS or not necessary?

File: src/maxtext/configs/types.py [hunk 6]
```diff
@@ -1158,6 +1129,26 @@ class GrainDataset(BaseModel):
...
+class OlmoGrainDataset(BaseModel):
+  """Configuration for the OLMo numpy fixed-seq-length input pipeline (dataset_type=olmo_grain).
+
+  Worker count, per-worker buffer size, and shuffle seed reuse the standard
+  grain flags (``grain_worker_count``, ``grain_per_worker_buffer_size``,
+  ``data_shuffle_seed``); only OLMo-specific fields are listed here.
+  """
+
+  olmo_index_path: PathStr = Field("", description="Path or gs:// URI to the JSON index from build_olmo_npy_index.py.")
+  olmo_path_remap_from: PathStr = Field(
```
Comment (RissyRan) [inline src/maxtext/configs/types.py:1177]: Maybe add some comments and indicate why we need this?

**Review by RissyRan — COMMENTED** [review]
Thanks for the change! 

Have you had a chance to chat with @aireenmei about this yet? I'm wondering if we could design these features to directly leverage the existing `grain_data_processing.py`. For example, we could add things like n-gram filtering and pre-tokenized numpy mixes as features there to improve code reusability.

The main benefit I see is reducing maintenance overhead. We currently maintain both tfds and c4_tfds_mlperf, but the latter is rarely used and has some maintenance issues. Since Grain will be heavily used moving forward, it makes sense to build on top of it directly. Let me know your thoughts—happy to discuss!

File: src/maxtext/input_pipeline/olmo_data.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+  true_locs = np.where(prog[:-1:] == prog[1::])[0]
+  return int(true_locs[0] + 1)
+
+
+def _find_start_last_consecutive_true(arr: np.ndarray) -> int:
+  """Start offset of the trailing run of True in ``arr``, or -1 if none."""
+  reverse = _find_end_first_consecutive_true(arr[::-1])
+  return len(arr) - reverse if reverse > 0 else -1
+
+
```
[the remark this answers is not in the document]
Comment (gagika) [reply src/maxtext/input_pipeline/olmo_data.py:342]: Done.

File: src/maxtext/input_pipeline/olmo_data.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+) -> Generator[RepetitionTuple, None, None]:
+  """Yield :class:`RepetitionTuple` for periodic spans of length ≥ 3 in
+  ``arr``. ``mask_value`` must not appear in the array (used internally as
+  reshape padding)."""
+  if (arr == mask_value).sum() > 0:
+    raise ValueError("`mask_value` is in the array")
+
+  max_period = min(max_period, len(arr) // 3)
+
+  for period in range(min_period, max_period + 1):
```
[the remark this answers is not in the document]
Comment (gagika) [reply src/maxtext/input_pipeline/olmo_data.py:367]: Updated the docstring: under uint32, -1 is the max uint32 value, way above any realistic vocab. Callers with vocabs that hit that id should pass an out-of-vocab sentinel.

File: docs/guides/data_input_pipeline/olmo_grain.md (hunk 5 above)
Comment (gagika) [reply docs/guides/data_input_pipeline/olmo_grain.md]: local path (updating the doc).

File: src/maxtext/configs/types.py (hunk 6 above)
Comment (gagika) [reply src/maxtext/configs/types.py:1177]: Done, added comments.

**Comment by gagika** [comment]
> Thanks for the change!
> 
> Have you had a chance to chat with @aireenmei about this yet? I'm wondering if we could design these features to directly leverage the existing `grain_data_processing.py`. For example, we could add things like n-gram filtering and pre-tokenized numpy mixes as features there to improve code reusability.
> 
> The main benefit I see is reducing maintenance overhead. We currently maintain both tfds and c4_tfds_mlperf, but the latter is rarely used and has some maintenance issues. Since Grain will be heavily used moving forward, it makes sense to build on top of it directly. Let me know your thoughts—happy to discuss!

I haven't yet chatted with @aireenmei, I will follow up.

The main reason I kept this separate is the stateless-resume contract — record at step k must be a pure function of `(seed, shard, k)` so we recover the data offset from the checkpoint step alone,  with no Grain iterator state in the checkpoint. `MapDataset.shuffle().repeat()` has no step-offset  hook, so reusing it as-is would drop that property.

The general pieces, e.g.  the n-gram mask, probably can be lifted into  `grain_data_processing.py` as a follow-up. Happy to take that on after this lands. 
I will chat with Aireen about it.

File: src/maxtext/input_pipeline/olmo_data_grain.py [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+
+  Checkpointing is trivial: the only mutable state is "which epoch's
+  permutation is currently cached" (a perf optimization). The user-visible
+  position is just the index passed to ``__getitem__``.
+
+  Args:
+    total_instances: ``index.total_instances`` from the OLMo index.
+    seed: Base seed for the shuffle.
+    shard_index: Zero-based index of this data-loading host. Typically
+      ``jax.process_index()``.
```
[the remark this answers is not in the document]
Comment (gagika) [reply src/maxtext/input_pipeline/olmo_data_grain.py:181]: at this scale it's fine, we can follow up with Chunked / Philox-keyed shuffle as follow-up if we need larger.

**Review by aireenmei — APPROVED** [review]

**Review by RissyRan — APPROVED** [review]
